### PR TITLE
Split cnn inspired by first layer of IMN

### PR DESCRIPTION
**Pull request by pmhalvor**
In reference to Issue #42, a split cnn functionality was implemented for fgflex.

Can be configured via params `split_cnn_kernels` and `split_cnn_tasks`.

Note: `split_cnn_tasks` only works when `split_cnn_kernels` is also provided.

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -345,8 +345,20 @@ def __init__(
...
 
+        # for i in range(5):  # dirty fix to avoid intermittend HTTPError for hugging face
+        #     try:
+        #         self.bert = BertModel.from_pretrained(bert_path).to(torch.device(self.device))
+        #     except Exception as e:
+        #         if "HTTPError" in e.msg() and i<4:
+        #             continue
+        #         raise e 
+        #     if self.find("bert") is not None:
+        #         break
```
Comment (pmhalvor) [inline src/model.py:356]: Could be an idea to build a safer loader form hugging face. Does not fail ofte, but annoying when it does.

File: src/model.py [hunk 2]
```diff
@@ -540,6 +552,56 @@ def expanding_cnn_block(self, in_channels, out_channels, kernel_size=5, m=2):
...
+            3. concatenate to original output size
+
+        Parameters:
+            kernels (list): sizes of kernels wished to split cnn between.
+                NOTE: kernel count must be divisible by 768, ex (1,2,3,4,6,8,...)
+        """
+        use_cnn_dropout = self.find("use_cnn_dropout", default=False)  # FIXME True?
+        use_cnn_activation = self.find("use_cnn_activation", default=False)
+
+        splits = torch.nn.ModuleList([])
```
Comment (pmhalvor) [inline src/model.py:570]: `ModuleList` for easier cast to device

File: src/model.py [hunk 3]
```diff
@@ -1927,29 +1992,70 @@ def init_components(self, subtasks):
...
         #######################################
-        for task in self.subtasks:
-            components[task].update(
-                {"cnn": torch.nn.Sequential(*[
-                    cnn_block(cnn_dim, cnn_dim)
-                    for layer in range(self.find(task+"_layers", default=1))
-                ])
-            })
+        # NOTE: stack can be 0, then these components are skipped. # TODO test this implementation
+        for stack in range(stack_count):
```
Comment (pmhalvor) [inline src/model.py:1996]: New components for every new stack, like RACL

File: src/model.py [hunk 4]
```diff
@@ -1927,29 +1992,70 @@ def init_components(self, subtasks):
...
+                            for layer in range(task_layers)
+                            ])}
+                        )
+                elif None not in (split_cnn_tasks, split_cnn_kernels): 
+                    if task in split_cnn_tasks:
+                        components[task].update({
+                            f"cnn_{stack}": self.split_cnn_block(
+                                cnn_dim, cnn_dim, split_cnn_kernels, task_layers
+                                )
+                            })
```
Comment (pmhalvor) [inline src/model.py:2014]: Both `split_cnn_kernels` and `split_cnn_tasks`  have to be present in order to add task wise split cnns. If they are not provided, then regular cnn blocks are used.

File: src/model.py [hunk 5]
```diff
@@ -1927,29 +1992,70 @@ def init_components(self, subtasks):
...
+        # NOTE main difference: IMN uses split_cnn_block w/ kernel=3 & kernel=5
+        if split_cnn_kernels is not None:  # NOTE only split on first shared
+            shared = torch.nn.ModuleDict({
+                "cnn_0": self.split_cnn_block(
+                    in_channels=768, 
+                    out_channels=cnn_dim, 
+                    kernels = split_cnn_kernels
+                ),
+                "attn_linear": self.linear_block(cnn_dim*2, 3)
+            })
```
Comment (pmhalvor) [inline src/model.py:2046]: Like IMN, apply split kernels to first layer of shared layers.

File: src/model.py [hunk 6]
```diff
@@ -1973,7 +2079,7 @@ def init_components(self, subtasks):
             "relations": relations,
         }))
 
-        if stack_count > 0:
+        if stack_count > 0:  # when stack == 0 no attention should be made. No task-wise cnn either.. 
             prev_task = None
             for task in self.subtasks:
                 for relation in components["relations"]["stack_0"]:
```
Comment (pmhalvor) [inline src/model.py:2085]: Is this block ever even reached?

File: src/model.py [hunk 7]
```diff
@@ -2019,41 +2127,67 @@ def forward(self, batch):
...
+            splits = len(shared[f"cnn_0"])
+            sentence_output = torch.cat(
+                [
+                    cnn(sentence_output)
+                    for cnn in shared[f"cnn_0"]
+                ],
+                dim=1  # concat on the embedding dimension
+            )
+        else:
+            sentence_output = shared[f"cnn_0"](sentence_output)
```
Comment (pmhalvor) [inline src/model.py:2142]: Handle shared split cnn.

File: src/model.py [hunk 8]
```diff
@@ -2106,7 +2242,9 @@ def forward(self, batch):
                 else:
                     same_task = [attn_inter]
 
-                task_inputs[first_task] = torch.stack(same_task).mean(dim=0)  
+                if first_task in task_inputs:
+                    re_encoded = self.components[first_task]["re_encode"](torch.stack(same_task).mean(dim=0))
+                    task_inputs[first_task] = re_encoded.permute(0, 2, 1)
```
Comment (pmhalvor) [inline src/model.py:2246]: Since the attention output is now used as the next task_input, we need to re-encode it to the orignal shared embedding size

File: src/plot.py [hunk 9]
```diff
@@ -97,6 +97,9 @@ def parse_metrics(data):
                 metric = metric.split("(RACL)")[-1].strip()
             loss = float(line.split('overall: ')[-1].split(' (')[0].strip())
             scores[metric.lower()].append(loss)
+        
+        if "Score:" in line:
+            print(line.split("INFO] ")[1])
```
Comment (pmhalvor) [inline src/plot.py:102]: Show final scores for each plot for better plot readability (in notebooks)

File: src/plot.py [hunk 10]
```diff
@@ -181,6 +184,23 @@ def show_same_smoothed(same, stop_at=1, header="smoothed", show_df=True):
...
+    avg_metric = sum([metric[i] for i in runs])/len(runs)
+
+    avg_loss.plot(title="Loss "+header, figsize=(15,5))
+    avg_metric.plot(title="Metric "+header, figsize=(15,5))
+
+    print("Final values:")
+    display(avg_loss.tail(3))
+    display(avg_metric.tail(3))
+
+    return avg_loss, avg_metric
```
Comment (pmhalvor) [inline src/plot.py:200]: For plotting multiple same runs (see issue #33)

Also displays the data frames (pretty print) of averages in notebook

File: src/utils.py [hunk 11]
```diff
@@ -351,7 +351,7 @@ def binary_tp(gold, pred):
     # for p in pred:  # for token in prediction
     tp = False  # reset tp value
     for p, g in zip(pred, gold):
-        if p == g:
+        if g > 0 and p == g:  # gold token is B or I and pred==gold
```
Comment (pmhalvor) [inline src/utils.py:354]: Binary F1 updated to only count trues when gold labels are not 0.

**Review by pmhalvor — COMMENTED** [review]
Working implementation currently testing on Fox (57324, 57325, 57328, 57329)

- [x] fix stack
- [x] implement split on shared 
- [x] implement split on tasks
- [ ] clean up TODOs

Some TODO still remain.
